homeassistant/components/homekit_controller/__init__.py

```diff
@@ -0,0 +1,218 @@
+                                'iid': service['iid']}
+                devtype = homekit.ServicesTypes.get_short(service['type'])
+                _LOGGER.debug("Found %s", devtype)
+                component = HOMEKIT_ACCESSORY_DISPATCH.get(devtype, None)
+                if component is not None:
+                    discovery.load_platform(self.hass, component, DOMAIN,
+                                            service_info, self.config)
+
+    def device_config_callback(self, callback_data):
+        """Handle initial pairing"""
+        import homekit
+        pairing_id = str(uuid.uuid4())
+        code = callback_data.get('code').strip()
+        self.pairing_data = homekit.perform_pair_setup(
+            self.conn, code, pairing_id)
+        if self.pairing_data is not None:
+            homekit.save_pairing(self.pairing_file, self.pairing_data)
+            self.accessory_setup()
+        else:
+            error_msg = "Unable to pair, please try again"
+            _configurator = self.hass.data[DOMAIN+self.hkid]
+            self.configurator.notify_errors(_configurator, error_msg)
+
+    def configure(self):
+        description = "Please enter the HomeKit code for your {}".format(
+            self.model)
+        self.hass.data[DOMAIN+self.hkid] = \
+            self.configurator.request_config(self.model,
+                                             self.device_config_callback,
+                                             description=description,
+                                             submit_caption="submit",
+                                             fields=[{'id': 'code',
+                                                      'name': 'HomeKit code',
+                                                      'type': 'string'}])
+
+
+class HomeKitEntity(Entity):
+    """Representation of a Home Assistant HomeKit device."""
+
+    def __init__(self, accessory, devinfo):
+        self._name = accessory.model
+        self._securecon = accessory.securecon
+        self._aid = devinfo['aid']
+        self._iid = devinfo['iid']
+        self._address = "homekit-{}-{}".format(devinfo['serial'], self._iid)
+        self._features = 0
+        self._chars = {}
+
+    def update(self):
+        response = self._securecon.get('/accessories')
+        data = json.loads(response.read().decode())
+        for accessory in data['accessories']:
+            if accessory['aid'] != self._aid:
+                continue
+            for service in accessory['services']:
+                if service['iid'] != self._iid:
+                    continue
+                self.update_characteristics(service['characteristics'])
+                break
+
+    @property
+    def unique_id(self):
+        """Return the ID of this device."""
+        return self._address
+
+    @property
+    def name(self):
+        """Return the name of the device if any."""
+        return self._name
+
+    def update_characteristics(self):
```

Comment:
Make sure this has the same parameters as the methods that you need to implement in the child classes, ie add `characteristics`.